index.js

```diff
@@ -3,57 +3,77 @@
 const fp = require('fastify-plugin')
 const formatEntry = require('./lib/formatEntry')
 const CRLF = '\r\n'
+const EarlyHints = `HTTP/1.1 103 Early Hint${CRLF}`
 
 function fastifyEarlyHints (fastify, opts, next) {
   if (fastify.initialConfig.http2 === true) {
-    return next(new Error('Early Hints cannot be used with a HTTP2 server.'))
+    return next(Error('Early Hints cannot be used with a HTTP2 server.'))
   }
 
   const formatEntryOpts = {
     warn: opts.warn
   }
 
-  function earlyHints (reply) {
-    const promiseBuffer = []
-    const serialize = function (c) {
-      let message = `HTTP/1.1 103 Early Hints${CRLF}`
-      for (let i = 0; i < c.length; i++) {
-        message += `${formatEntry(c[i], formatEntryOpts)}${CRLF}`
-      }
-      return `${message}${CRLF}`
-    }
-
-    return {
-      close: async function () {
-        if (promiseBuffer.length) {
-          await Promise.all(promiseBuffer)
+  fastify.decorateReply('writeEarlyHints', function (headers) {
+    const reply = this
+    let message = ''
+    if (Array.isArray(headers)) {
+      for (const nameValues of headers) {
+        if (typeof nameValues === 'object' && typeof nameValues.name === 'string' && typeof nameValues.value === 'string') {
+          message += `${nameValues.name}: ${nameValues.value}${CRLF}`
+        } else {
+          throw Error('"headers" expected to be name-value object')
         }
-      },
-      add: function (content) {
-        const p = new Promise(resolve => {
-          if (reply.raw.socket) {
-            reply.raw.socket.write(serialize(content), 'utf-8', resolve)
-          } else {
-            reply.raw.write(serialize(content), 'utf-8', resolve)
+      }
+    } else if (typeof headers === 'object' && headers !== null) {
+      for (const key of Object.keys(headers)) {
+        if (Array.isArray(headers[key])) {
+          for (const value of headers[key]) {
+            message += `${key}: ${value}${CRLF}`
           }
-        })
-        promiseBuffer.push(p)
-        return p
+        } else {
+          message += `${key}: ${headers[key]}${CRLF}`
+        }
       }
+    } else {
+      throw Error(`"headers" expected to be object or Array, but recieved ${typeof headers}`)
     }
-  }
 
-  function onRequest (request, reply, done) {
-    reply.eh = earlyHints(reply)
-    done()
-  }
+    return new Promise(function (resolve) {
+      if (reply.raw.socket === null) {
+        resolve()
+        return
+      }
+      reply.raw.socket.write(`${EarlyHints}${message}${CRLF}`, 'ascii', () => {
+        // we do not care the message is sent or lost. Since early hints
+        // is metadata to instructs the clients to do action before actual
```

Comment:
```suggestion
        // is metadata to instruct the clients to do something before actual
```

not sure about "something", but "to do action" doesn't read right either

Reply:
It is up to the client to decide, it can be `prefetch`, `ignore` or anything else.